pkg/kv/kvserver/rafttrace/rafttrace.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package rafttrace
+
+import (
+	"context"
+	"fmt"
+	"sync/atomic"
+
+	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
+	"github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvserverpb"
+	"github.com/cockroachdb/cockroach/pkg/raft"
+	"github.com/cockroachdb/cockroach/pkg/raft/raftpb"
+	"github.com/cockroachdb/cockroach/pkg/settings"
+	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/logtags"
+)
+
+// traceValue represents the trace information for a single registration.
+type traceValue struct {
+	traceID tracingpb.TraceID
+	spanID  tracingpb.SpanID
+	// ctx is a trace specific context used to log events on this trace.
+	ctx context.Context
+	// baseCtx is the underlying proposal buffer which we additionally log to.
+	baseCtx withContext
+	// appRespSent and storageAppendRespSent track whether range messages have
+	// already been logged to log at most once. This limits the log from growing
+	// too large at a small risk of missing some messages in the case of dropped
+	// messages or reproposals.
```

Comment:
Good idea. We could also have a "watermark" in the `RaftTracer` instead of storing a bool in every `traceValue`, but as is is also fine.

Reply:
I wanted to keep the individual traces independent and I noticed that during heavy load the number of dupe logs was pretty high. Hopefully this balances things well.